packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts

```diff
@@ -144,6 +147,59 @@ export class ActorExtractLinksSolidTypeIndex extends ActorExtractLinks {
     return typeLinks;
   }
 
+  /**
+   * To fetch the domain of the predicate.
+   * @param predicateSubjects A dictionary of predicate and its subjects from the query.
+   * @param typeSubjects A dictionary of class type and its subjects from the query.
+   */
+  public async linkPredicateDomains(predicateSubjects: Record<string, RDF.Term>,
+    typeSubjects: Record<string, RDF.Term[]>): Promise<void> {
+    if (Object.keys(predicateSubjects).length > 0) {
+      const predicateDomainsInner = await Promise.all(Object.keys(predicateSubjects)
+        .map(predicate => this.fetchPredicateDomains(predicate)));
+      const predicateDomainsRec = Object.fromEntries(predicateDomainsInner);
+      for (const [ predicate, subject ] of Object.entries(predicateSubjects)) {
+        const typeNames = predicateDomainsRec[predicate];
+        if (typeNames) {
+          for (const typeName of typeNames) {
+            if (!typeSubjects[typeName]) {
+              typeSubjects[typeName] = [];
+            }
+            typeSubjects[typeName].push(subject);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * To fetch the rdf type from the vocabulary if the type is not already present.
+   * @param predicateValue Predicate value from the query.
+   * @return predicateTypeLinks A record mapping predicate URIs to it's domain.
+   */
+  public async fetchPredicateDomains(predicateValue: IDataSource): Promise<(string | string[])[]> {
+    const bindings = await this.queryEngine.queryBindings(`
+        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+        SELECT * WHERE {
+          ?s rdfs:domain ?domain.
+        }`, {
+      sources: [ predicateValue ],
+      [KeysRdfResolveHypermediaLinks.traverse.name]: false,
+    });
+
+    const bindingsArray = await bindings.toArray();
+    const predicateTypeLinks: (string | string[])[] = [];
+    const domainsArray: string[] = [];
+    let predicate = '';
```

Comment:
Why was this code changed? The previous version looked fine to me. Not sure this one is fully correct.